Content.Client/GameObjects/Components/Instruments/InstrumentComponent.cs

```diff
@@ -266,29 +286,71 @@ public bool CloseMidi()
         /// <param name="midiEvent">The received midi event</param>
         private void RendererOnMidiEvent(MidiEvent midiEvent)
         {
-            _midiQueue.Enqueue(midiEvent);
+            _midiQueue.Add(midiEvent);
         }
 
-        public override void Update(float delta)
-        {
-            if (!IsMidiOpen && !IsInputOpen)
-                return;
 
-            _timer -= delta;
+        private TimeSpan _lastMeasuredInterval = TimeSpan.MinValue;
 
-            if (_timer > 0f) return;
+        private int _sentWithinASec = 0;
 
-            SendAllMidiMessages();
-            _timer = TimeBetweenNetMessages;
-        }
+        private static readonly TimeSpan OneSecAgo = TimeSpan.FromSeconds(-1);
 
-        private void SendAllMidiMessages()
+        private readonly Dictionary<byte, TimeSpan> _notesActive = new Dictionary<byte,TimeSpan>();
+
+        public override void Update(float delta)
         {
+            if (!IsMidiOpen && !IsInputOpen) return;
+
+            var now = _gameTiming.RealTime;
+            var oneSecAGo = now.Add(OneSecAgo);
+
+            if (_lastMeasuredInterval <= oneSecAGo)
+            {
+                _lastMeasuredInterval = now;
+                _sentWithinASec = 0;
+            }
+
             if (_midiQueue.Count == 0) return;
-            var events = _midiQueue.ToArray();
-            _midiQueue.Clear();
+
+            var max = Math.Min(MaxMidiEventsPerBatch, MaxMidiEventsPerSecond - _sentWithinASec);
+
+            if (max <= 0)
+            {
+                _midiSawmill.Info($"Playing {_notesActive.Count} notes, hit {_sentWithinASec} event/sec limit.");
```

Comment:
Remove all this logging as it will spam the console.

Reply:
Maybe after it gets tested.

Reply:
Ah fuck it.